pyat/at/errors/error_def.py

```diff
@@ -0,0 +1,449 @@
+        rr = ring.replace(refpts)
+        for e in rr[refpts]:
+            setattr(e, pname, get_pol(e, pname, pstat, pdyn, index))
+            sanitize(e)
+        return rr
+
+    alldef = kwargs.pop('all', True)
+    for pol in ['A', 'B']:
+        if kwargs.pop('Polynom'+pol, alldef):
+            index = kwargs.pop('Index'+pol, None)
+            ring = set_polerr(ring, 'Polynom'+pol, index)
+    return ring
+    
+
+def enable_errors(ring: Lattice, **kwargs):
+    r"""Enable magnet errors
+
+    Returns a shallow copy of *ring* where magnet errors defined by
+    :py:func:`assign_errors` are introduced in the magnet attributes.
+
+    By default, all assigned errors are enabled, but keyword arguments allow a
+    selection of the enabled errors.
+
+    Args:
+        ring:   Lattice description.
+
+    Keyword Args:
+        all (bool):         Set the default value for all the specific error
+          flags. Default: :py:obj:`True`
+        shiftx (bool):      enable horizontal shift errors. Default: *all*
+        shifty (bool):      enable vertical shift errors. Default: *all*
+        tilt (bool):        enable tilt errors. Default: *all*
+        pitch (bool):       enable pitch errors. Default: *all*
+        yaw (bool):         enable yaw errors. Default: *all*
+        PolynomB (bool):    enable polynomial errors. Default: *all*
+        IndexB (Optional[int]): restrict the polynomial errors to the
+          specified index. Default: :py:obj:`None` meaning the whole polynom
+        PolynomA (bool):    enable polynomial errors. Default: *all*
+        IndexA (Optional[int]): restrict the skew polynomial error to the
+          specified index. Default: :py:obj:`None` meaning the whole polynom
+
+    Examples:
+        >>> ringerr = enable_errors(ring)
+
+        All errors are enabled in *ringerr*
+
+        >>> ringerr = enable_errors(all=False, ShiftErr=True)
+
+        Only magnet shift errors are active
+
+        >>> ringerr = enable_errors(all=False)
+
+        No magnet errors are enabled. Only monitor errors will be active.
+
+    See also:
+        :py:func:`assign_errors`, :py:func:`get_optics_err`
+    """
+    if not ring.has_errors:
+        ring = _apply_field_errors(ring, **kwargs)
+        ring = _apply_alignment_errors(ring, **kwargs)
+        ring._has_errors = True
+    return ring
+
+
+def has_errors(ring):
+    """Get the error status of a lattice"""
+    return getattr(ring, '_has_errors', False)
+
+
+def get_mean_std_err(ring: Lattice, key, attr, index=0):
+    vals = [np.atleast_1d(getattr(e, attr, 0.0))[index]
+            for e in ring.get_elements(key)]
+    return np.mean(vals), np.std(vals)
+
+
+def _sort_flags(kwargs):
+    errargs = {}
+    for key in _SEL_ARGS:
+        if key in kwargs:
+            errargs[key] = kwargs.pop(key)
+    return kwargs, errargs
+
+
+def find_orbit_err(ring: Lattice, refpts: Refpts = None, **kwargs):
```

Comment:
an additional issue to be dealt by `find_orbit` are BPM misalignments (T1/T2 R1/R2) for my personal usage I have always considered this to be the position of the BPM in the machine. A different information compared to the BPM offset/rotation. In a lattice with very large misalignment, (potentially following a smooth curve with amplitudes reaching several millimeters), T1/T2 could be order of mm, offsets of order of micrometers and the final corrected orbit also micrometers. To read the "absolute" orbit in space, one would have to sum T2 to the BPM readings. However this information is almost never relevant, it may be needed just for a single plot to show what find_orbit returns. It may be provided via an absolute=False flag. If True T2 are ignored.
`find_orbit` make indirectly use of the T1-2 in all elements to give eventually a very large orbit. 
T1 and T2 errors are ignored by `_apply_bpm_orbit_errors` but it may be needed to include them in order to give the user the real expected measured orbit with errors.

Reply:
At the moment, T1-2 and R1-2 in `Monitors` don't have any effect on anything: they don't affect the orbit, neither with nor without errors. So you can indeed use them for your personal use to keep track of any information you want. On the other hand, if you have a mechanical offset of several millimetres, you can put it in the `BpmShift` attribute (beware: the sign is currently debated with @lcarver), you may get a closed orbit of micrometers, or even zero if there are no magnet errors, but the BPM reading as returned by `find_orbit_err` will be millimetres: that's correct !

Reply:
I tend to agree this is a bit tricky, how do you handle the case where T1/T2 are non zero in the elements upstream and downstream the BPM?